kotlinx-collections-immutable/src/main/kotlin/kotlinx/collections/immutable/ImmutableCollection.kt

```diff
@@ -16,24 +16,76 @@
 
 package kotlinx.collections.immutable
 
+/**
+ * A generic immutable collection of elements. Methods in this interface support only read-only access to the collection;
+ * modification operations are supported through the [PersistentCollection] interface.
+ *
+ * Implementors of this interface take responsibility to be immutable,
+ * that is objects reachable from an instance of the implementor should never be mutated.
+ *
+ * @param E the type of elements contained in the collection. The immutable collection is covariant on its element type.
+ */
 public interface ImmutableCollection<out E>: Collection<E>
 
+/**
+ * A generic persistent collection of elements that supports adding and removing elements.
+ * Modification operations return new instances of the persistent collection with the modification applied.
+ *
+ * @param E the type of elements contained in the collection. The persistent collection is covariant on its element type.
+ */
 public interface PersistentCollection<out E> : ImmutableCollection<E> {
+    /**
+     * Returns a new instance of the persistent collection with the specified [element] added.
+     */
     fun add(element: @UnsafeVariance E): PersistentCollection<E>
 
+    /**
+     * Returns a new instance of the persistent collection with the specified elements added.
+     */
     fun addAll(elements: Collection<@UnsafeVariance E>): PersistentCollection<E>
 
+    /**
+     * Returns a new instance of the persistent collection with a single instance of the specified [element],
+     * if it is present, removed.
+     */
     fun remove(element: @UnsafeVariance E): PersistentCollection<E>
 
+    /**
+     * Returns a new instance of the persistent collection with all elements in this collection that are also
+     * contained in the specified collection [elements] removed.
+     */
     fun removeAll(elements: Collection<@UnsafeVariance E>): PersistentCollection<E>
 
+    /**
+     * Returns a new instance of the persistent collection with all elements in this collection that match
+     * the specified [predicate] removed.
+     */
     fun removeAll(predicate: (E) -> Boolean): PersistentCollection<E>
 
+    /**
+     * Returns an empty instance of the persistent collection.
+     */
     fun clear(): PersistentCollection<E>
 
+    /**
+     * A generic builder of the persistent collection.
+     * Builder exposes its modification operations through the [MutableCollection] interface.
+     * Builders are reusable, that is [build] method can be called multiple times with modifications between these calls.
+     * However, modifications applied should not affect previously built persistent collection instances.
+     */
     interface Builder<E>: MutableCollection<E> {
+        /**
+         * Returns a new instance of the persistent collection containing all elements of this builder.
+         * It is allowed to call this method multiple times.
```

Comment:
We can also document that if no modification were applied, it returns the same persistent collection instance it was obtained from.

Reply:
+1